designs/2022-unused-disable-directive-flexible-config/README.md

```diff
@@ -0,0 +1,247 @@
+- `lib/options.js` - switch to string / number and use new default
+- Anywhere `reportUnusedDisableDirectives` is passed around as a boolean needs to change to using a severity level (and no more need to convert between boolean and severity level)
+- A few test files that set `reportUnusedDisableDirectives` as a boolean will be updated to use a severity level
+
+Luckily, `reportUnusedDisableDirectives` is already stored as a severity level in much of the underlying code.
+
+## Documentation
+
+<!--
+    How will this RFC be documented? Does it need a formal announcement
+    on the ESLint blog to explain the motivation?
+-->
+
+These documentation pages will be updated to reflect the new behavior of these options:
+
+- Configuration Files
+- Configuring Rules
+- Command Line Interface
+- Node.js API
+- `eslint --help`
+
+## Drawbacks
+
+<!--
+    Why should we *not* do this? Consider why adding this into ESLint
+    might not benefit the project or the community. Attempt to think
+    about any opposing viewpoints that reviewers might bring up.
+
+    Any change has potential downsides, including increased maintenance
+    burden, incompatibility with other tools, breaking existing user
+    experience, etc. Try to identify as many potential problems with
+    implementing this RFC as possible.
+-->
+
+- [Breaking changes](#backwards-compatibility-analysis) cause churn and disruption.
+- Potential [semver policy implications](#semver-policy) as described below.
+
+### Semver Policy
+
+Erroring by default on unused disable directives will require slight tweaks to our [semantic versioning policy](https://github.com/eslint/eslint#semantic-versioning-policy). It currently reads:
+
+> Patch release (intended to not break your lint build)
+> > A bug fix in a rule that results in ESLint reporting fewer linting errors.
+> > ...
+>
+> Minor release (might break your lint build)
+> > A bug fix in a rule that results in ESLint reporting more linting errors.
+> > ...
+
+These lines will need to be adjusted because a bug fix that causes a rule to produce fewer errors will now break builds if disable directives become unused as a result. See this context:
+
+- <https://github.com/eslint/eslint/issues/12703#issuecomment-568582014>
+  > If we allowed for errors to be reported with reportUnusedDisableDirectives, we would limit what kind of changes we could publish in a semver-patch release (since fixing a bug that would result in fewer errors could create additional unused disable directives).
+- <https://github.com/eslint/eslint/pull/14699#discussion_r650863268>
+  > ...because of our semver policy, which says that bug fixes that produce fewer errors should not break builds.
+
+#### Semver Policy Alternative 1
+
+One possibility for updating the semver policy is that we could consider any bug fix to a rule, regardless of whether it results in more or less violations, as simply a bug fix and therefore suitable for release as a patch release.
```

Comment:
If we end up defaulting unused disable directives to `warn` (being discussed in https://github.com/eslint/rfcs/pull/100/files#r1044770688), then I guess this discussion of changing the semver policy will become moot.

---

Regarding this requirement:

> The rationale is that we don't want to break people's CI systems. If their CI system is passing on a Friday and we do a patch release over the weekend, their CI system should still be passing on Monday when they arrive at the office.

I do want to note that "Semver Policy Alternative 2" would be compatible with this requirement that patch releases not break lint builds (any changes to the number of violations would be released as a minor release instead):

> Semver Policy Alternative 2: If we want to maintain the categorization of changes based on whether they can break your lint build, then any change to the number of violations a rule produces would have to change from being patch release compatible to only minor release compatible.

---

Regarding "Semver Policy Alternative 1":

I would find it helpful to my own understanding if you or others could elaborate on the disagreements with my specific supporting points for Alternative 1:

> Semver Policy Alternative 1: Consider any bug fix to a rule, regardless of whether it results in more or less violations, as simply a bug fix and therefore suitable for release as a patch release.
>
> - It's how I believe most of the ecosystem (i.e. ESLint plugins) treats bug fixes in real-world usage.
> - It's also similar to [Prettier's policy](https://github.com/prettier/prettier/issues/2201) that allows formatting tweaks/fixes in any type of release:
>   > If prettier produces a different output on two different versions, that's not a breaking change. Ideally prettier will never have any breaking changes, but if there are, it would be to the the API/CLI itself, not the output.
> - Some bug fixes cause a rule to produce a combination of more violations and less violations. The distinction between the two is not always clear-cut and not necessarily deserving of different treatment in the semver policy.
> - From what I can tell, attempts to work around the existing semver policy are why we have ended up with such a convoluted and unsatisfying design for reporting unused disable directives in the first place, so it could be time to rethink this constraint.
> - In addition, the current restriction of lint-build-breaking changes to minor releases already seems arbitrary, as technically, any breaking changes should be limited to major releases according to semver. This obviously wouldn't be practical for us, as it would prevent us from making any bug fixes to rules except in annual major releases. So since ESLint already does not strictly follow semver in terms of breaking lint builds, why impose this arbitrary limitation on what types of bug fixes can be made in patch releases?
> - Even today, it's somewhat aspirational to proclaim that a patch release will never break your lint build.

Reply:
I'm sorry, I don't really feel like it's a good use of time or energy to go point-for-point with each of your arguments. Suffice to say, there have been a lot of discussion about semver policy and a lot of reasons why we ended up where we did. If you want to learn more, I'd suggest searching through the tsc-meetings repo where you can find both the meeting notes and chat transcripts for past meetings.

Reply:
Okay. I have removed the change to the semver policy.